src/openvswitch.c

```diff
@@ -0,0 +1,146 @@
+        // LCOV_EXCL_STOP
+    }
+}
+
+#define OPENVSWITCH_OVS_VSCTL "/usr/bin/ovs-vsctl"
+#define append_systemd_cmd(s, command, ...) \
+{ \
+    g_string_append(s, "ExecStart="); \
+    g_string_append_printf(s, command, __VA_ARGS__); \
+    g_string_append(s, "\n"); \
+}
+
+static char*
+netplan_type_to_table_name(const NetplanDefType type)
+{
+    switch (type) {
+        case NETPLAN_DEF_TYPE_BRIDGE:
+            return "Bridge";
+        default:
+            return "Interface";
+    }
+}
+
+static void
+write_ovs_additional_data(GHashTable *data, const NetplanDefType type, const gchar* id_escaped, GString* cmds, const char* setting)
+{
+    GHashTableIter iter;
+    gchar* key;
+    gchar* value;
+
+    g_hash_table_iter_init(&iter, data);
+    while (g_hash_table_iter_next(&iter, (gpointer) &key, (gpointer) &value)) {
+        append_systemd_cmd(cmds, OPENVSWITCH_OVS_VSCTL " set %s %s %s:%s=%s",
+                           netplan_type_to_table_name(type), id_escaped, setting, key, value);
+    }
+}
+
+/**
+ * Generate the OpenVSwitch systemd units for configuration of the selected netdef
+ * @rootdir: If not %NULL, generate configuration in this root directory
+ *           (useful for testing).
+ */
+void
+write_ovs_conf(const NetplanNetDefinition* def, const char* rootdir)
+{
+    GString* cmds = g_string_new(NULL);
+    g_autofree gchar* id_escaped = NULL; // TODO: Check if this works
+
+    id_escaped = systemd_escape(def->id);
+
+    /* TODO: error out on non-existing ovs-vsctl tool */
+    /* TODO: maybe dynamically query the ovs-vsctl tool path? */
+
+    /* Common OVS settings can be specified even for non-OVS interfaces */
+    if (def->ovs_settings.external_ids && g_hash_table_size(def->ovs_settings.external_ids) > 0) {
+        write_ovs_additional_data(def->ovs_settings.external_ids, def->type,
+                                          id_escaped, cmds, "external-ids");
+    }
+
+    if (def->ovs_settings.other_config && g_hash_table_size(def->ovs_settings.other_config) > 0) {
+        write_ovs_additional_data(def->ovs_settings.other_config, def->type,
+                                          id_escaped, cmds, "other-config");
+    }
+
+    /* For other, more OVS specific settings, we expect the backend to be set to OVS */
+    if (def->backend == NETPLAN_BACKEND_OVS) {
+        /* TODO */
+    } else {
+        g_debug("openvswitch: definition %s is not for us (backend %i)", def->id, def->backend);
+    }
+
+    /* If we need to configure anything for this netdef, write the required systemd unit */
+    if (cmds->len > 0)
+        write_ovs_systemd_unit(id_escaped, cmds, rootdir);
+    g_string_free(cmds, TRUE);
+}
+
+/**
+ * Finalize the OpenVSwitch configuration (global config)
+ */
+void
+write_ovs_conf_finish(const char* rootdir)
+{
+    /* TODO: Write the global config here */
+}
+
+/**
+ * Clean up all generated configurations in @rootdir from previous runs.
+ */
+void
+cleanup_ovs_conf(const char* rootdir)
```

Comment:
We need to make sure to correctly clean up everything here.

And setup the correct dependencies in `netplan/cli/commands/apply.py`, to avoid the recent errors we have seen with failing `netplan-wpa-IFNAME.service` units.

Reply:
Ah, yes! Some of the cleanup was here but I forgot to uncomment it - I commented it out when I was debugging if everything works as expected. Uncommenting.